daemon/restart.go

```diff
@@ -35,11 +35,20 @@ func (daemon *Daemon) containerRestart(container *container.Container, seconds i
 		defer daemon.Unmount(container)
 	}
 
-	if err := daemon.containerStop(container, seconds); err != nil {
+	// set AutoRemove flag to false before stop so the container won't be
+	// removed during restart process
+	autoRemove := container.HostConfig.AutoRemove
+
+	container.HostConfig.AutoRemove = false
+	err := daemon.containerStop(container, seconds)
+	// restore AutoRemove after container is stopped successfully
```

Comment:
I think this should be:
`restore AutoRemove irrespective of whether the stop worked or not`